File: cmd/agent_mode.go
package cmd

import (
	"os"
	"strings"

	"github.com/spf13/cobra"
)

var autoAgentDetectionEnvVars = []string{
	"AI_AGENT",
	"AGENT",
	"CLAUDECODE",
	"CLAUDE_CODE",
	"CLAUDE_CODE_ENTRYPOINT",
	"CURSOR_AGENT",
	"CURSOR_TRACE_ID",
	"CURSOR_EXTENSION_HOST_ROLE",
	"CODEX_SANDBOX",
	"CODEX_CI",
	"CODEX_THREAD_ID",
	"AIDER",
	"AIDER_MODE",
	"CLINE_ACTIVE",
	"WINDSURF",
	"WINDSURF_MODE",
	"GITHUB_COPILOT",
	"COPILOT_MODE",
	"AMAZON_Q",
	"Q_AGENT",
	"GEMINI_CLI",
	"GEMINI_CODE_ASSIST",
	"CODY",
}

func agentModeEnabled(cmd *cobra.Command) bool {
	if cmd != nil {
		if flag := cmd.Flag("agent-mode"); flag != nil && flag.Changed {
			return agentModeFlag
		}
	}

	return detectAgentEnvironment()
}

func detectAgentEnvironment() bool {
	return detectAgentEnvironmentFromLookup(os.LookupEnv)
}

func detectAgentEnvironmentFromLookup(lookup func(string) (string, bool)) bool {
	for _, envVar := range autoAgentDetectionEnvVars {
		value, ok := lookup(envVar)
		if !ok {
			continue
		}
		if agentSignalEnabled(value) {
			return true
		}
	}

	return false
}

func agentSignalEnabled(value string) bool {
	normalized := strings.ToLower(strings.TrimSpace(value))
	switch normalized {
	case "", "0", "false", "no", "off", "none", "disabled":
		return false
	default:
		return true
	}
}

File: src/agent-mode.ts
export const autoAgentDetectionEnvVars = [
  "AI_AGENT",
  "AGENT",
  "CLAUDECODE",
  "CLAUDE_CODE",
  "CLAUDE_CODE_ENTRYPOINT",
  "CURSOR_AGENT",
  "CURSOR_TRACE_ID",
  "CURSOR_EXTENSION_HOST_ROLE",
  "CODEX_SANDBOX",
  "CODEX_CI",
  "CODEX_THREAD_ID",
  "AIDER",
  "AIDER_MODE",
  "CLINE_ACTIVE",
  "WINDSURF",
  "WINDSURF_MODE",
  "GITHUB_COPILOT",
  "COPILOT_MODE",
  "AMAZON_Q",
  "Q_AGENT",
  "GEMINI_CLI",
  "GEMINI_CODE_ASSIST",
  "CODY",
] as const;

function agentSignalEnabled(value: string): boolean {
  const normalized = value.trim().toLowerCase();
  switch (normalized) {
    case "":
    case "0":
    case "false":
    case "no":
    case "off":
    case "none":
    case "disabled":
      return false;
    default:
      return true;
  }
}

export function detectAgentEnvironment(
  lookup: (name: string) => string | undefined = (name) => process.env[name],
): boolean {
  for (const envVar of autoAgentDetectionEnvVars) {
    const value = lookup(envVar);
    if (value !== undefined && agentSignalEnabled(value)) {
      return true;
    }
  }
  return false;
}

export function agentModeEnabled(explicit?: boolean): boolean {
  if (explicit !== undefined) {
    return explicit;
  }
  return detectAgentEnvironment();
}
